server/src/main/java/org/elasticsearch/cluster/metadata/DataStream.java

```diff
@@ -408,8 +414,72 @@ public Settings getEffectiveSettings(ProjectMetadata projectMetadata) {
+     * rolled over right now. This includes the mapping overrides, the mapping from the composable template, and the mappings from all
+     * component templates in the composable template.
+     * @param projectMetadata
+     * @param composableTemplate The composable template that matches the data stream's name
+     * @param mappingsOverrides  The mapping overrides to be applied
+     * @param writeIndex The write index of the data stream whose effective mappings are to be returned. This is used only to determine
+     *                   whether data stream mappings ought to be added when calling collectV2Mappings, and to create an IndexService
+     *                   object. The mappings that will be used come from templateWithMergedMappings and from the component templates, not
+     *                   from this writeIndex.
+     * @param indicesService
+     * @return A JSON CompressedXContent representation of the effective mappings for the composableTemplate plus mappingsOverrides
+     * @throws IOException
+     */
+    @SuppressWarnings("unchecked")
+    public static CompressedXContent getEffectiveMappings(
+        ProjectMetadata projectMetadata,
+        ComposableIndexTemplate composableTemplate,
+        CompressedXContent mappingsOverrides,
+        Index writeIndex,
+        IndicesService indicesService
+    ) throws IOException {
+        ComposableIndexTemplate mergedTemplate = composableTemplate.mergeMappings(mappingsOverrides);
+        final String requestMappings = null; // We are not using any request mappings
+        final NamedXContentRegistry xContentRegistry = null; // This is only used to parse requestMappings if they are not null
+        final List<CompressedXContent> mappings = collectV2Mappings(
+            requestMappings,
+            projectMetadata,
+            mergedTemplate,
+            xContentRegistry,
+            writeIndex.getName()
+        );
+        return indicesService.withTempIndexService(projectMetadata.index(writeIndex), indexService -> {
+            MapperService mapperService = indexService.mapperService();
+            DocumentMapper documentMapper = mapperService.merge(
+                MapperService.SINGLE_MAPPING_NAME,
+                mappings,
+                MapperService.MergeReason.INDEX_TEMPLATE
+            );
+            Map<String, ?> resultMapping;
+            Map<String, Object> originalMappingMap = XContentHelper.convertToMap(
+                documentMapper.mappingSource().uncompressed(),
+                true,
+                XContentType.JSON
+            ).v2();
+            if (originalMappingMap.containsKey(MapperService.SINGLE_MAPPING_NAME)) {
+                resultMapping = (Map<String, ?>) originalMappingMap.get(MapperService.SINGLE_MAPPING_NAME);
+            } else {
+                resultMapping = originalMappingMap;
+            }
+            return convertMappingMapToXContent(resultMapping);
```

Comment:
Seems a little strange that we uncompress the xcontent, convert it to a map, potentially strip the mapping name, then reconvert it to a compressed xcontent. Should we instead leave the mapping name so that we don't have to do the extra conversions?

Reply:
I don't understand the question. The default is `_doc`, which is `MapperService.SINGLE_MAPPING_NAME`, right?

Reply:
I believe we usually just deal with it if it exists. It seems confusing as a user (given that most users are probably unaware that we used to allow multiples since that has been gone for quite a while). I had just been trying to make it simpler for users of a new API so that they didn't have to write client code that can handle the result two different ways.